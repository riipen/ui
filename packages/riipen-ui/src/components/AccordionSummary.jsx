import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";
import css from "styled-jsx/css";

import ThemeContext from "../styles/ThemeContext";

import AccordionContext from "./AccordionContext";
import ButtonIcon from "./ButtonIcon";

const AccordionSummary = props => {
  const { children, classes, iconProps, icon, ...other } = props;

  const theme = React.useContext(ThemeContext);

  const [focusedState, setFocusedState] = React.useState(false);

  const getLinkedStyles = () => {
    return css.resolve`
      .root {
        align-items: center;
        cursor: pointer;
        display: flex;
        transition: ${theme.transitions.create([
          "background-color",
          "min-height"
        ])};
        user-select: none;
      }
      .root.disabled {
        cursor: initial;
      }

      .content {
        flex-grow: 1;
        transition: ${theme.transitions.create(["margin"])};
      }

      .icon {
        transform: rotate(0deg);
        transition: ${theme.transitions.create(["transform"])};
      }
      .icon.expanded {
        transform: rotate(180deg);
      }
    `;
  };

  const handleFocusVisible = () => {
    setFocusedState(true);
  };

  const handleBlur = () => {
    setFocusedState(false);
  };

  const { disabled = false, expanded, toggle } = React.useContext(
    AccordionContext
  );

  const handleClick = event => {
    if (toggle && !disabled) {
      toggle(event);
    }
  };

  const linkedStyles = getLinkedStyles();

  return (
    <div
      disabled={disabled}
      aria-expanded={expanded}
      className={clsx(
        linkedStyles.className,
        "root",
        {
          disabled,
          expanded,
          focused: focusedState
        },
        classes
      )}
      onClick={handleClick}
      onFocusVisible={handleFocusVisible}
      onBlur={handleBlur}
      {...other}
    >
      <div
        className={clsx(linkedStyles.className, "content", {
          expanded
        })}
      >
        {children}
      </div>
      {icon && (
        <ButtonIcon
          component="div"
          {...iconProps}
          classes={[
            clsx(
              linkedStyles.className,
              "icon",
              {
                expanded
              },
              iconProps.classes
            )
          ]}
          disabled={disabled}
        >
          {icon}
        </ButtonIcon>
      )}
      {linkedStyles.styles}
    </div>
  );
};

AccordionSummary.propTypes = {
  /**
   * The content of the accordion summary.
   */
  children: PropTypes.node,

  /**
   * List of additional classes to apply to this component.
   */
  classes: PropTypes.array,

  /**
   * The icon to display as the expand indicator.
   */
  icon: PropTypes.node,

  /**
   * Props applied to the `ButtonIcon` element wrapping the expand icon.
   */
  iconProps: PropTypes.object
};

AccordionSummary.defaultProps = {
  classes: [],
  iconProps: {}
};

AccordionSummary.displayName = "AccordionSummary";

export default AccordionSummary;
