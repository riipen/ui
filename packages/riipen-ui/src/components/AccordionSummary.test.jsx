import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronDown } from "@fortawesome/free-solid-svg-icons";

import AccordionSummary from "./AccordionSummary";

const Icon = <FontAwesomeIcon icon={faChevronDown} />;

describe("<AccordionSummary>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<AccordionSummary />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<AccordionSummary />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("sets correct default props", () => {
    const wrapper = mount(<AccordionSummary />);

    expect(wrapper.find("AccordionSummary").props().classes).toEqual([]);
    expect(wrapper.find("AccordionSummary").props().iconProps).toEqual({});
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = <div>A child</div>;

      const wrapper = mount(<AccordionSummary>{child}</AccordionSummary>);

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classes = ["newClass"];

      const wrapper = mount(<AccordionSummary classes={classes} />);

      expect(
        wrapper
          .find("AccordionSummary")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });
  });

  describe("icon prop", () => {
    it("renders icon given valid custom icon prop", () => {
      const icon = Icon;

      const wrapper = mount(<AccordionSummary icon={icon} />);

      expect(wrapper.find(".icon").type()).toEqual("div");
    });
  });

  describe("iconProps prop", () => {
    it("passes given iconProp attributes to icon prop", () => {
      const icon = Icon;
      const iconProps = { size: "small" };

      const wrapper = mount(
        <AccordionSummary icon={icon} iconProps={iconProps} />
      );

      expect(wrapper.find("ButtonIcon").props().size).toEqual(iconProps.size);
    });
  });
});
