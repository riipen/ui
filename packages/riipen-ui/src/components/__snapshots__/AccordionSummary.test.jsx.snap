// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccordionSummary> renders correct snapshot 1`] = `
<AccordionSummary
  classes={Array []}
  iconProps={Object {}}
>
  <div
    className="jsx-3687637527 root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocusVisible={[Function]}
  >
    <div
      className="jsx-3687637527 content"
    />
    <JSXStyle
      dynamic={
        Array [
          "background-color 300ms cubic-bezier(0.4, 0, 0.2, 1), min-height 300ms cubic-bezier(0.4, 0, 0.2, 1)",
          "margin 300ms cubic-bezier(0.4, 0, 0.2, 1)",
          "transform 300ms cubic-bezier(0.4, 0, 0.2, 1)",
        ]
      }
      id="1644211938"
    />
  </div>
</AccordionSummary>
`;
